### PR TITLE
[Refactor] Upgrade ray from 1.9.1 to 2.1.0

### PR DESCRIPTION
**Pull request by KKIEEK**
Related to #41

## Sub PRs
* https://github.com/SIAnalytics/siatune/pull/91
* https://github.com/SIAnalytics/siatune/pull/97 
* https://github.com/SIAnalytics/siatune/pull/100 
* https://github.com/SIAnalytics/siatune/pull/102 
* https://github.com/SIAnalytics/siatune/pull/104

### Todo
- [ ] Support MMEngine
- [ ] Supplement documentations

### CODE REVIEW

**Review by yhna940 — APPROVED** [review]
LGTM

File: siatune/mm/tasks/base.py [hunk 1]
```diff
@@ -29,9 +29,7 @@ class BaseTask(metaclass=ABCMeta):
             1. args (argparse.Namespace): The low level CLI arguments.
             2. searched_cfg (Dict):
                 The configuration searched by the algorithm.
-            3. checkpoint_dir (Optional[str]):
```
Comment (yhna940) [inline siatune/mm/tasks/base.py:32]: Edited the description.

File: siatune/apis/analysis.py [hunk 2]
```diff
@@ -5,19 +5,19 @@
...
-def log_analysis(analysis: tune.ExperimentAnalysis,
+def log_analysis(results: ResultGrid,
                  tune_config: Config,
                  task_config: Optional[Config] = None,
                  log_dir: Optional[str] = None) -> None:
     """Log the analysis of the experiment.
 
     Args:
-        analysis (tune.ExperimentAnalysis): The analysis of the experiment.
+        results (ResultGrid): Result of `Tuner.fit()`.
```
Comment (nijkah) [inline siatune/apis/analysis.py:20]:
```suggestion
        results (ResultGrid): Experiment results of `Tuner.fit()`.
```

File: siatune/mm/tasks/mmtrainbase.py [hunk 3]
```diff
@@ -56,52 +54,22 @@ def train_model(
...
+    def create_trainable(self) -> TorchTrainer:
         """Get ray trainable task.
 
         Args:
-            backend (str):
-                The backend for dist training. Defaults to 'nccl'.
-            timeout_s (int):
-                Seconds before the torch process group times out.
+            backend (str): The backend for distributed training.
+                Defaults to 'nccl'.
```
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:62]:
```suggestion
```

File: siatune/mm/tasks/mmtrainbase.py [hunk 4]
```diff
@@ -56,52 +54,22 @@ def train_model(
...
-        if backend == 'nccl' and os.getenv('NCCL_BLOCKING_WAIT') is None:
-            os.environ['NCCL_BLOCKING_WAIT'] = '0'
-        return super().context_aware_run(searched_cfg, **kwargs)
-
-    def create_trainable(
-        self,
-        backend: str = 'nccl',
-        timeout_s: int = 1800,
-    ) -> ray.tune.trainable:
+    def create_trainable(self) -> TorchTrainer:
```
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:57]:
```suggestion
    def create_trainable(self, backend='nccl') -> TorchTrainer:
```

File: siatune/mm/tasks/mmtrainbase.py [hunk 5]
```diff
@@ -56,52 +54,22 @@ def train_model(
...
-        )
+        return TorchTrainer(
+            self.context_aware_run,
+            scaling_config=ScalingConfig(
+                trainer_resources=dict(
+                    CPU=self.num_cpus_per_worker,
+                    GPU=self.num_gpus_per_worker),
+                num_workers=self.num_workers,
+                use_gpu=torch.cuda.is_available()),
+            torch_config=TorchConfig(backend='gloo'))
```
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:75]:
```suggestion
            torch_config=TorchConfig(backend=backend))
```

**Review by nijkah — DISMISSED** [review]

File: siatune/ray/tuner.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+            param_space=dict(train_loop_config=param_space),
+            tune_config=TuneConfig(
+                search_alg=searcher, scheduler=trial_scheduler, **tune_cfg),
+            run_config=RunConfig(
+                local_dir=work_dir,
+                stop=stopper,
+                callbacks=callbacks,
+                failure_config=None,  # todo
+                sync_config=None,  # todo
+                checkpoint_config=None,  # todo
```
Comment (nijkah) [inline siatune/ray/tuner.py:71]: Are these required to be edited in this PR? or just future works?

File: siatune/mm/tasks/mmtrainbase.py (hunk 4 above)
Comment (KKIEEK) [reply siatune/mm/tasks/mmtrainbase.py:57]: The `backend` argument is no longer required in `create_trainable ` method.

File: siatune/ray/tuner.py [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+            trial_scheduler=cfg.get('trial_scheduler', None),
+            stopper=cfg.get('stopper', None),
+            callbacks=cfg.get('callbacks', None),
+        )
+
+        return tuner
+
+    @classmethod
+    def resume(cls, path, **kwargs):
+        return cls.restore(path, **kwargs)
```
Comment (nijkah) [inline siatune/ray/tuner.py:93]: Is this working?

Comment (KKIEEK) [reply siatune/ray/tuner.py:93]: Not yet. We need to work more.

File: siatune/mm/tasks/mmtrainbase.py (hunk 4 above)
Comment (nijkah) [reply siatune/mm/tasks/mmtrainbase.py:57]: Then, the backend for `TorchConfig` is always fixed to `'gloo'`?

File: siatune/mm/tasks/mmtrainbase.py (hunk 5 above)
Comment (KKIEEK) [reply siatune/mm/tasks/mmtrainbase.py:75]: Please refer to [this PR](https://github.com/SIAnalytics/siatune/pull/91)

File: siatune/ray/tuner.py (hunk 6 above)
Comment (KKIEEK) [reply siatune/ray/tuner.py:71]: I'm thinking

File: siatune/mm/tasks/mmtrainbase.py (hunk 4 above)
Comment (KKIEEK) [reply siatune/mm/tasks/mmtrainbase.py:57]: The `backend` argument is determined by the `init_process_group` in `run` method.

File: siatune/mm/tasks/base.py (hunk 1 above)
Comment (KKIEEK) [reply siatune/mm/tasks/base.py:32]: This line was deleted.

File: siatune/mm/tasks/mmtrainbase.py (hunk 3 above)
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:62]:
```suggestion
```

File: siatune/mm/tasks/mmtrainbase.py [hunk 8]
```diff
@@ -58,8 +58,6 @@ def create_trainable(self) -> TorchTrainer:
         """Get ray trainable task.
 
         Args:
```
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:60]:
```suggestion
```

File: siatune/mm/tasks/mmtrainbase.py [hunk 9]
```diff
@@ -56,52 +55,21 @@ def train_model(
...
 
-        Args:
-            backend (str):
-                The backend for dist training. Defaults to 'nccl'.
-            timeout_s (int):
-                Seconds before the torch process group times out.
-
         Returns:
-            ray.tune.trainable: The trainable task.
+            TorchTrainer: The trainable task.
```
Comment (nijkah) [inline siatune/mm/tasks/mmtrainbase.py:62]:
```suggestion
            DataParallelTrainer: The trainable task.
```

File: siatune/ray/tuner.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+        searcher (dict, optional): Search algorithm for optimization.
+            Default to random search.
+            Refer to :module:`ray.tune.search` for more options.
+        trial_scheduler (dict, optional): Scheduler for executing the trial.
+            Default to FIFO scheduler.
+            Refer to :module:`ray.tune.schedulers` for more options.
+        stopper (dict, optional): Stop conditions to consider.
+            Refer to :class:`ray.tune.stopper.Stopper` for more info.
+        callbacks (dict | list, optional): Callbacks to invoke.
+            Refer to :class:``ray.tune.callback.Callback` for more info.
```
Comment (KKIEEK) [inline siatune/ray/tuner.py:34]:
```suggestion
            Refer to :class:`ray.tune.callback.Callback` for more info.
```

File: siatune/ray/tuner.py (hunk 7 above)
Comment (KKIEEK) [reply siatune/ray/tuner.py:102]: Now the `resume` method works! #104

**Review by nijkah — APPROVED** [review]

**Pull request merged**